File: 21. Merge Two Sorted Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null) return l2;
        if (l2 == null) return l1;
        
        ListNode dummy = new ListNode(0);
        ListNode l3 = dummy;
        
        while(l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                l3.next = l1;
                l1 = l1.next;
            }
            else {
                l3.next = l2;
                l2 = l2.next;
            }
            l3 = l3.next;
        }
        
        if (l1 == null && l2 != null) l3.next = l2;
        if (l1 != null && l2 == null) l3.next = l1;
        
        return dummy.next;
        
//         ArrayList<Integer> items = new ArrayList<>();
        
//         while(l1 != null) {
//             items.add(l1.val);
//             l1 = l1.next;
//         }
        
//         while(l2 != null) {
//             items.add(l2.val);
//             l2 = l2.next;
//         }
        
//         Collections.sort(items);
        
        
//         ListNode first = new ListNode(0);
//         ListNode head = first;
        
//         for(int value : items) {
//             head.next = new ListNode(value);
//             head = head.next;
//         }
//         return first.next;
    }
}

File: amazon/21. Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head = ListNode(0)
        dummy = head
        
        if (not l1) and (not l2):
            return None
        
        while l1 or l2:
            if l1 and l2:
                val1 = l1.val 
                val2 = l2.val
                if val1 < val2:
                    dummy.next = l1
                    l1 = l1.next
                else:
                    dummy.next = l2
                    l2 = l2.next
            elif l1: # l2 is None
                dummy.next = l1
                break
            elif l2: # l1 is None
                dummy.next = l2
                break
            dummy = dummy.next
        
        return head.next
            
